Fix done message parse

zhaofeng0019 opened the pull request:
Though the document of kernel (https://kernel.org/doc/html/next/userspace-api/netlink/intro.html#netlink-message-types) specify the format of netlink message with NLMSG_DONE, it also says that "Note that some implementations may issue custom NLMSG_DONE messages in reply to do action requests. In that case the payload is implementation-specific and may also be absent.".

After searching the source code of kernel, we can find that
1. the formant specified in the document is obeyed by most generic netlink but some generic netlink like this (https://elixir.bootlin.com/linux/v6.15/source/drivers/net/team/team_core.c#L2494) has no payload, so as a generic lib, we should not suppose the format of DoneMessage. it's sensible to just save the payload.

2. when use NLMSG_DONE as an end of multi messages, there will always be a NLM_F_MULTIPART in the flag and only in this case should we parse it as a DoneMessage, in other occassion like connector netlink (https://elixir.bootlin.com/linux/v6.15/source/drivers/connector/connector.c#L101), we should parse it as a common message.

zhaofeng0019 commented:
> @zhaofeng0019 Thanks for the detailed explanation!
> 
> Please give me sometime to test this PR as this might impact all other `rust-netlink` crates on handling DONE message.
> 
> Meanwhile:
> 
> 1. Please fix the `cargo fmt` CI failure.
> 2. Please provide example on kernel code using `custom NLMSG_DONE messages` instead of 0 size blob.

1. fixed
2. it seems that in current kernel tree, custom NLMSG_DONE messages instead of 0 size blob are all without NLM_F_MULTIPART, which can be covered by flag check

so what's your opinion? maybe a DoneBuffer which can endure empty buffer  to avoid break change? It's fine to me. 
BTW，some third-party kernel module use netlink to communicate can still use every format they wnat to mark  an end for mulit messages.

zhaofeng0019 commented:
i found this DoneMessage issue when i was using cn_proc to monitor process following the sample code https://elixir.bootlin.com/linux/v6.15/source/tools/testing/selftests/connector/proc_filter.c#L51, then i found out all message i received had been parse to DoneMessage with a code and ext_act but not the format i want like this https://elixir.bootlin.com/linux/v6.15/source/include/uapi/linux/connector.h#L70. You can try if you like.

zhaofeng0019 commented:
> I am a `no-issue-no-change` guy. Do we have any actual issue for current code or just your imaginary based on kernel document and code?
> 
> Of course, if you are coding a proprietary kernel module, no need to share the detail.

try this
```rust
use std::{error::Error, fmt};

use netlink_packet_core::{
    NLMSG_DONE, NetlinkDeserializable, NetlinkHeader, NetlinkMessage, NetlinkPayload,
    NetlinkSerializable,
};

use netlink_sys::{Socket, SocketAddr, protocols::NETLINK_CONNECTOR};

#[derive(Debug, Clone)]
pub struct CnMessage {
    pub idx: u32,
    pub val: u32,
    pub seq: u32,
    pub ack: u32,
    pub len: u16,
    pub flags: u16,
    pub data: Vec<u8>,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct DeserializeError(&'static str);
impl Error for DeserializeError {
    fn description(&self) -> &str {
        self.0
    }
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        None
    }
}

impl fmt::Display for DeserializeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl NetlinkDeserializable for CnMessage {
    type Error = DeserializeError;

    fn deserialize(_header: &NetlinkHeader, payload: &[u8]) -> Result<Self, Self::Error> {
        let idx = u32::from_ne_bytes(payload[0..4].try_into().unwrap());
        let val = u32::from_ne_bytes(payload[4..8].try_into().unwrap());
        let seq = u32::from_ne_bytes(payload[8..12].try_into().unwrap());
        let ack = u32::from_ne_bytes(payload[12..16].try_into().unwrap());
        let len = u16::from_ne_bytes(payload[16..18].try_into().unwrap());
        let flags = u16::from_ne_bytes(payload[18..20].try_into().unwrap());
        let date = payload[20..(20 + len as usize)].to_vec();
        Ok(Self {
            idx,
            val,
            seq,
            ack,
            len,
            flags,
            data: date,
        })
    }
}

impl NetlinkSerializable for CnMessage {
    fn message_type(&self) -> u16 {
        NLMSG_DONE
    }

    fn buffer_len(&self) -> usize {
        20 + self.data.len()
    }

    fn serialize(&self, buffer: &mut [u8]) {
        buffer[0..4].copy_from_slice(&self.idx.to_ne_bytes());
        buffer[4..8].copy_from_slice(&self.val.to_ne_bytes());
        buffer[8..12].copy_from_slice(&self.seq.to_ne_bytes());
        buffer[12..16].copy_from_slice(&self.ack.to_ne_bytes());
        buffer[16..18].copy_from_slice(&self.len.to_ne_bytes());
        buffer[18..20].copy_from_slice(&self.flags.to_ne_bytes());
        buffer[20..(20 + self.data.len())].copy_from_slice(&self.data);
    }
}

fn test_cn_proc() {
    let mut socket = Socket::new(NETLINK_CONNECTOR).unwrap();
    socket.bind(&SocketAddr::new(0, 1)).unwrap();
    let cn_msg_req = CnMessage {
        idx: 1,
        val: 1,
        len: 4,
        seq: 1,
        ack: 1,
        flags: 1,
        data: vec![1, 0, 0, 0],
    };
    let mut request = NetlinkMessage::new(
        NetlinkHeader::default(),
        NetlinkPayload::InnerMessage(cn_msg_req),
    );
    request.finalize();
    let mut buff = vec![0; request.buffer_len()];
    request.serialize(&mut buff);
    socket.send(&buff, 0).unwrap();
    let mut recv_buff = Vec::with_capacity(128);
    let l = socket.recv(&mut recv_buff, 0).unwrap();
    let message = NetlinkMessage::<CnMessage>::deserialize(&recv_buff[0..l]).unwrap();
    println!("{message:?}");

    let cn_msg_req = CnMessage {
        idx: 1,
        val: 1,
        len: 4,
        seq: 1,
        ack: 1,
        flags: 1,
        data: vec![2, 0, 0, 0],
    };
    let mut request = NetlinkMessage::new(
        NetlinkHeader::default(),
        NetlinkPayload::InnerMessage(cn_msg_req),
    );
    request.finalize();
    let mut buff = vec![0; request.buffer_len()];
    request.serialize(&mut buff);
    socket.send(&buff, 0).unwrap();
}

fn main() {
    test_cn_proc();
}

```

zhaofeng0019 commented:
> I am a `no-issue-no-change` guy. Do we have any actual issue for current code or just your imaginary based on kernel document and code?
> 
> 
> 
> Of course, if you are coding a proprietary kernel module, no need to share the detail.

if you like, i can still give you an example that a DoneMessage without payload thus fail to be parsed. it's easy to write according to the kernel code.

zhaofeng0019 commented:
anyone to take a look？

cathay4t commented:
Sorry for the delay. Working on it.

cathay4t commented:
Aha, kernel has `nlmsg_put(skb, portid, seq, NLMSG_DONE, 0, flags | NLM_F_MULTI);`

Trying in real world

cathay4t commented:
Instead of dramatic changes to API, please use this patch:

```patch
diff --git a/src/done.rs b/src/done.rs
index f81507c..6dee1ac 100644
--- a/src/done.rs
+++ b/src/done.rs
@@ -9,7 +9,7 @@ use crate::{Emitable, Field, Parseable, Rest};
 
 const CODE: Field = 0..4;
 const EXTENDED_ACK: Rest = 4..;
-const DONE_HEADER_LEN: usize = EXTENDED_ACK.start;
+pub(crate) const DONE_HEADER_LEN: usize = EXTENDED_ACK.start;
 
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[non_exhaustive]
diff --git a/src/message.rs b/src/message.rs
index 4bc7dda..eae70f9 100644
--- a/src/message.rs
+++ b/src/message.rs
@@ -6,6 +6,7 @@ use anyhow::Context;
 use netlink_packet_utils::DecodeError;
 
 use crate::{
+    done::DONE_HEADER_LEN,
     payload::{NLMSG_DONE, NLMSG_ERROR, NLMSG_NOOP, NLMSG_OVERRUN},
     DoneBuffer, DoneMessage, Emitable, ErrorBuffer, ErrorMessage,
     NetlinkBuffer, NetlinkDeserializable, NetlinkHeader, NetlinkPayload,
@@ -105,9 +106,15 @@ where
             }
             NLMSG_NOOP => Noop,
             NLMSG_DONE => {
-                let msg = DoneBuffer::new_checked(&bytes)
-                    .and_then(|buf| DoneMessage::parse(&buf))
-                    .context("failed to parse NLMSG_DONE")?;
+                let msg = if bytes.len() == 0 {
+                    DoneBuffer::new_checked(&[0u8; DONE_HEADER_LEN])
+                        .and_then(|buf| DoneMessage::parse(&buf))
+                        .context("failed to parse NLMSG_DONE")?
+                } else {
+                    DoneBuffer::new_checked(&bytes)
+                        .and_then(|buf| DoneMessage::parse(&buf))
+                        .context("failed to parse NLMSG_DONE")?
+                };
                 Done(msg)
             }
             NLMSG_OVERRUN => Overrun(bytes.to_vec()),
```

cathay4t commented:
I have create PR https://github.com/rust-netlink/netlink-packet-core/pull/32

xiaoff0019 committed:
Fix done message parse
Though the document of kernel (https://kernel.org/doc/html/next/userspace-api/netlink/intro.html#netlink-message-types)
specify the format of netlink message with NLMSG_DONE, it also says that "Note that some implementations may issue custom NLMSG_DONE messages in reply to do action requests.
In that case the payload is implementation-specific and may also be absent.".
when use NLMSG_DONE as an end of multi messages, there will always be a NLM_F_MULTIPART in the flag and only in this case should we parse it as a DoneMessage, in other occassion like connector netlink
(https://elixir.bootlin.com/linux/v6.15/source/drivers/connector/connector.c#L101), we should parse it as a common message.

Signed-off-by: feng zhao <zhaofeng0019@gmail.com>

zhaofeng0019 commented:
new commit has been made

cathay4t commented:
Oh, I am in the middle of deploying new version `netlink-packet-core 0.8` into other rust-netlink crates, your code might not able to compile yet.

cathay4t commented:
Please use latest netlink_packet_generic and netlink_packet_core to try again.

zhaofeng0019 commented:
i have tried, but the latest pkg doesn't handle connector netlink correctly

cathay4t commented:
Sorry I missed your message.

Any detail on `i have tried, but the latest pkg doesn't handle connector netlink correctly`?
